Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public protocol BufferStorage<Element> {
+  associatedtype Element
+  @Sendable func send(element: Element) async
```

Comment:
could you elaborate a bit please ?

Reply:
so in generic version you have:
```swift
@rethrows
public protocol BufferStorage<Element>: Sendable {
  associatedtype Element
  func send(element: Element) async
  func fail(error: Error)
  func finish()
  func next() async -> Result<Element?, Error>
}
```

Adding throws to the `send` method allows for sending to the buffer to fail; e.g. throwing when the buffer is full.

```swift
@rethrows
public protocol BufferStorage<Element>: Sendable {
  associatedtype Element
  func send(element: Element) async throws
  func fail(error: Error)
  func finish()
  func next() async -> Result<Element?, Error>
}
```

Reply:
ok I see. I'll try to add that requirement. It should ha 0 impact on my code I think since the thrown exception will be catch within the `Task` and transmitted to the buffer as an error. It will make the `next` function throw when the consumer consumers that error.

As for this signature `func next() async throws -> Element?`, I'll give it a try.

I'm currently reworking a lit the code base. I have merged the `func finish()` and `func fail(error: Error)` functions into a single one `func finish(error: Error?)` and I've reworked the state machines to have less different states.

I guess I should have something by tomorrow now.

Reply:
I've changed the function signatures to match your suggestion.

Reply:
Overall I'm happy with the result.
The perfs are good (+20% for the unbounded algorithm compared to the Actor version).
The `AsyncStorage` protocol requirements are simple.

What do you think ?